File: typescript/src/lookupIndices.ts
import { getEntries, getCorrections } from "./data";
import { NameEntry } from "./types";

let built = false;

const arIndex: Map<string, NameEntry> = new Map();
const enIndex: Map<string, NameEntry> = new Map();
const arNormIndex: Map<string, NameEntry> = new Map();
const correctionIndex: Map<string, string> = new Map();
let ranked: NameEntry[] = [];
let allEntries: NameEntry[] = [];

// Regex for Arabic normalization
const ALEF_VARIANTS = /[\u0622\u0623\u0625\u0671]/g;
const TASHKEEL = /[\u0610-\u061A\u064B-\u065F\u0670\u06D6-\u06ED]/g;
const TATWEEL = /\u0640/g;
const ALEF_MAQSURA = "\u0649";
const YA = "\u064A";
const TA_MARBUTA = "\u0629";
const HA = "\u0647";

export function normalizeAr(text: string): string {
  let s = text.replace(TASHKEEL, "");
  s = s.replace(TATWEEL, "");
  s = s.replace(ALEF_VARIANTS, "\u0627");
  s = s.replace(new RegExp(ALEF_MAQSURA, "g"), YA);
  s = s.replace(new RegExp(TA_MARBUTA, "g"), HA);
  return s;
}

export function normalizeEn(text: string): string {
  return text.toLowerCase().replace(/-/g, "").replace(/'/g, "").trim();
}

function claimEn(key: string, entry: NameEntry): void {
  const existing = enIndex.get(key);
  if (!existing || entry.corpusShare > existing.corpusShare) {
    enIndex.set(key, entry);
  }
}

/**
 * Bind an Arabic variant spelling to the lemma with the larger corpus
 * share, same rule as `claimEn`.
 *
 * A canonical key (some entry's own `ar`/normalized `ar`) always wins
 * over any OTHER entry's variant claiming the same string — a rare
 * misspelling must never shadow a real lemma's own canonical spelling.
 * Among two variants with no canonical claim, the higher corpus share
 * wins, exactly like `claimEn`.
 */
function claimArVariant(
  index: Map<string, NameEntry>,
  canonicalKeys: Set<string>,
  key: string,
  entry: NameEntry
): void {
  if (canonicalKeys.has(key)) {
    // Already bound to its own entry in the canonical pass; a variant
    // from a different lemma must never override it.
    return;
  }
  const existing = index.get(key);
  if (!existing || entry.corpusShare > existing.corpusShare) {
    index.set(key, entry);
  }
}

function build(): void {
  if (built) return;

  const entries = getEntries();
  const corrections = getCorrections();

  // AR index, pass 1: canonical spellings are unconditional and take
  // priority over any other lemma's variant claiming the same string
  // (book has zero duplicate canonical ar values).
  const canonicalArKeys = new Set(entries.map((e) => e.ar));
  const canonicalArNormKeys = new Set(entries.map((e) => normalizeAr(e.ar)));
  for (const entry of entries) {
    arIndex.set(entry.ar, entry);
    arNormIndex.set(normalizeAr(entry.ar), entry);
  }

  for (const entry of entries) {
    // AR index, pass 2: variants. Keep the higher-share lemma when two
    // rows' variants claim the same spelling — same rule as English
    // keys, so a rare misspelling cannot steal a common name's lookup.
    for (const v of entry.arVariants) {
      const vStripped = v.trim();
      if (vStripped) {
        claimArVariant(arIndex, canonicalArKeys, vStripped, entry);
        claimArVariant(arNormIndex, canonicalArNormKeys, normalizeAr(vStripped), entry);
      }
    }

    // EN index — keep the higher-share lemma on a colliding key
    claimEn(normalizeEn(entry.en), entry);
    for (const v of entry.enVariants) {
      const vStripped = v.trim();
      if (vStripped) {
        claimEn(normalizeEn(vStripped), entry);
      }
    }
  }

  // Correction index
  for (const [k, v] of Object.entries(corrections)) {
    correctionIndex.set(k, v);
  }

  allEntries = [...entries];
  ranked = [...entries].sort((a, b) => b.corpusShare - a.corpusShare);
  built = true;
}

function ensureBuilt(): void {
  if (!built) {
    build();
  }
}

export function lookupAr(name: string): NameEntry | undefined {
  ensureBuilt();
  if (!name || !name.trim()) return undefined;
  const trimmed = name.trim();

  // 1. Direct match
  const direct = arIndex.get(trimmed);
  if (direct) return direct;

  // 2. Normalized match
  const norm = normalizeAr(trimmed);
  const normMatch = arNormIndex.get(norm);
  if (normMatch) return normMatch;

  // 3. Alif / Alif Maqsura terminal phonetic equivalence (e.g. مصطفا <-> مصطفى, موسا <-> موسى)
  if (norm.endsWith("\u0627")) {
    const alt = norm.slice(0, -1) + "\u064A";
    const altMatch = arNormIndex.get(alt);
    if (altMatch) return altMatch;
  } else if (norm.endsWith("\u064A")) {
    const alt = norm.slice(0, -1) + "\u0627";
    const altMatch = arNormIndex.get(alt);
    if (altMatch) return altMatch;
  }

  // 4. Space-less compound match (e.g. عبد الرحيم <-> عبدالرحيم)
  const noSpace = trimmed.replace(/\s+/g, "");
  if (noSpace !== trimmed) {
    const noSpaceMatch = arIndex.get(noSpace) || arNormIndex.get(normalizeAr(noSpace));
    if (noSpaceMatch) return noSpaceMatch;
  }

  return undefined;
}

export function lookupEn(name: string): NameEntry | undefined {
  ensureBuilt();
  return enIndex.get(normalizeEn(name));
}

export function isArabic(text: string): boolean {
  return /[\u0600-\u06FF\uFE70-\uFEFF]/.test(text);
}

export function lookup(name: string): NameEntry | undefined {
  ensureBuilt();
  if (isArabic(name)) {
    return lookupAr(name);
  }
  return lookupEn(name);
}

export function getCorrect(surface: string): string | undefined {
  ensureBuilt();
  return correctionIndex.get(surface);
}

export function getRanked(): NameEntry[] {
  ensureBuilt();
  return ranked;
}

export function getAll(): NameEntry[] {
  ensureBuilt();
  return allEntries;
}

export function getArForms(): Map<string, NameEntry> {
  ensureBuilt();
  return arIndex;
}

export function getArNormForms(): Map<string, NameEntry> {
  ensureBuilt();
  return arNormIndex;
}

File: dart/egy_names/lib/src/lookup_indices.dart
import 'data.dart';
import 'types.dart';

class LookupIndices {
  static bool _built = false;

  static final Map<String, NameEntry> _arIndex = {};
  static final Map<String, NameEntry> _enIndex = {};
  static final Map<String, NameEntry> _arNormIndex = {};
  static final Map<String, String> _correctionIndex = {};
  static List<NameEntry> _allEntries = [];
  static List<NameEntry> _rankedEntries = [];

  static final RegExp _tashkeelRegex =
      RegExp(r'[\u0610-\u061A\u064B-\u065F\u0670\u06D6-\u06ED]');
  static final RegExp _tatweelRegex = RegExp(r'\u0640');
  static final RegExp _alefVariantsRegex = RegExp(r'[\u0622\u0623\u0625\u0671]');
  static final RegExp _isArabicRegex =
      RegExp(r'[\u0600-\u06FF\uFE70-\uFEFF]');

  static String normalizeAr(String text) {
    var s = text.replaceAll(_tashkeelRegex, '');
    s = s.replaceAll(_tatweelRegex, '');
    s = s.replaceAll(_alefVariantsRegex, '\u0627');
    s = s.replaceAll('\u0649', '\u064A'); // ى -> ي
    s = s.replaceAll('\u0629', '\u0647'); // ة -> ه
    return s;
  }

  static String normalizeEn(String text) {
    return text.toLowerCase().replaceAll('-', '').replaceAll("'", '').trim();
  }

  static void _claimEn(String key, NameEntry entry) {
    final existing = _enIndex[key];
    if (existing == null || entry.corpusShare > existing.corpusShare) {
      _enIndex[key] = entry;
    }
  }

  /// Bind an Arabic variant spelling to the lemma with the larger
  /// corpus share, same rule as English keys.
  ///
  /// A canonical key (some entry's own `ar`/normalized `ar`) always
  /// wins over any OTHER entry's variant claiming the same string — a
  /// rare misspelling must never shadow a real lemma's own canonical
  /// spelling. Among two variants with no canonical claim, the higher
  /// corpus share wins, exactly like [_claimEn].
  static void _claimArVariant(
    Map<String, NameEntry> index,
    Set<String> canonicalKeys,
    String key,
    NameEntry entry,
  ) {
    if (canonicalKeys.contains(key)) {
      // Already bound to its own entry in the canonical pass; a
      // variant from a different lemma must never override it.
      return;
    }
    final existing = index[key];
    if (existing == null || entry.corpusShare > existing.corpusShare) {
      index[key] = entry;
    }
  }

  static bool isArabic(String text) {
    return _isArabicRegex.hasMatch(text);
  }

  static void ensureBuilt({String? dataPath}) {
    if (_built) return;

    final bundle = DataLoader.loadBundle(customPath: dataPath);
    _allEntries = List.unmodifiable(bundle.names);

    // AR index, pass 1: canonical spellings are unconditional and take
    // priority over any other lemma's variant claiming the same string
    // (the book has zero duplicate canonical ar values).
    final canonicalArKeys = _allEntries.map((e) => e.ar).toSet();
    final canonicalArNormKeys =
        _allEntries.map((e) => normalizeAr(e.ar)).toSet();
    for (final entry in _allEntries) {
      _arIndex[entry.ar] = entry;
      _arNormIndex[normalizeAr(entry.ar)] = entry;
    }

    for (final entry in _allEntries) {
      // AR index, pass 2: variants. Keep the higher-share lemma when
      // two rows' variants claim the same spelling — same rule as
      // English keys, so a rare misspelling cannot steal a common
      // name's lookup.
      for (final v in entry.arVariants) {
        final stripped = v.trim();
        if (stripped.isNotEmpty) {
          _claimArVariant(_arIndex, canonicalArKeys, stripped, entry);
          _claimArVariant(
            _arNormIndex,
            canonicalArNormKeys,
            normalizeAr(stripped),
            entry,
          );
        }
      }

      // EN Index — keep the higher-share lemma on a colliding key
      _claimEn(normalizeEn(entry.en), entry);

      for (final v in entry.enVariants) {
        final stripped = v.trim();
        if (stripped.isNotEmpty) {
          _claimEn(normalizeEn(stripped), entry);
        }
      }
    }

    // Corrections
    _correctionIndex.addAll(bundle.corrections);

    // Ranked
    final sorted = List<NameEntry>.from(_allEntries);
    sorted.sort((a, b) => b.corpusShare.compareTo(a.corpusShare));
    _rankedEntries = List.unmodifiable(sorted);

    _built = true;
  }

  static NameEntry? lookupAr(String name, {String? dataPath}) {
    ensureBuilt(dataPath: dataPath);
    if (name.trim().isEmpty) return null;
    final trimmed = name.trim();

    // 1. Direct match
    final direct = _arIndex[trimmed];
    if (direct != null) return direct;

    // 2. Normalized match
    final norm = normalizeAr(trimmed);
    final normMatch = _arNormIndex[norm];
    if (normMatch != null) return normMatch;

    // 3. Alif / Alif Maqsura terminal phonetic equivalence (e.g. مصطفا <-> مصطفى)
    if (norm.endsWith('\u0627')) {
      final alt = '${norm.substring(0, norm.length - 1)}\u064A';
      final altMatch = _arNormIndex[alt];
      if (altMatch != null) return altMatch;
    } else if (norm.endsWith('\u064A')) {
      final alt = '${norm.substring(0, norm.length - 1)}\u0627';
      final altMatch = _arNormIndex[alt];
      if (altMatch != null) return altMatch;
    }

    // 4. Space-less compound match (e.g. عبد الرحيم <-> عبدالرحيم)
    final noSpace = trimmed.replaceAll(RegExp(r'\s+'), '');
    if (noSpace != trimmed) {
      final noSpaceMatch = _arIndex[noSpace] ?? _arNormIndex[normalizeAr(noSpace)];
      if (noSpaceMatch != null) return noSpaceMatch;
    }

    return null;
  }

  static NameEntry? lookupEn(String name, {String? dataPath}) {
    ensureBuilt(dataPath: dataPath);
    return _enIndex[normalizeEn(name)];
  }

  static NameEntry? lookup(String name, {String? dataPath}) {
    ensureBuilt(dataPath: dataPath);
    if (isArabic(name)) {
      return lookupAr(name, dataPath: dataPath);
    }
    return lookupEn(name, dataPath: dataPath);
  }

  static String? getCorrection(String surface, {String? dataPath}) {
    ensureBuilt(dataPath: dataPath);
    return _correctionIndex[surface];
  }

  static List<NameEntry> getAll({String? dataPath}) {
    ensureBuilt(dataPath: dataPath);
    return _allEntries;
  }

  static List<NameEntry> getRanked({String? dataPath}) {
    ensureBuilt(dataPath: dataPath);
    return _rankedEntries;
  }

  static Map<String, NameEntry> getArForms({String? dataPath}) {
    ensureBuilt(dataPath: dataPath);
    return _arIndex;
  }

  static Map<String, NameEntry> getArNormForms({String? dataPath}) {
    ensureBuilt(dataPath: dataPath);
    return _arNormIndex;
  }
}
